### PR TITLE
Add basic skeleton for developer docs

### PR DESCRIPTION
**Pull request by systemsoverload**
Create a reasonable skeleton for important topics to cover and write a bit about the various subjects. This mostly note-taking levels of detail, but I think a reasonable starting off point. I'm open to any and all suggestions from copy-edits to missing/incomplete/incorrect data.

### CODE REVIEW

**Comment by ollpu** [comment]
Thanks! This is quite comprehensive already and the structure makes sense. I'm about halfway through reviewing.

On writing style, some parts feel sales-pitch-y in a way that I don't think is relevant in developer documentation. There are some inaccuracies that can be adjusted but also some nonsensical statements. (Did you use an LLM to help writing? That's fine, but I'd appreciate transparency on that)

**Comment by systemsoverload** [comment]
Yeah the outline was generated with my dev notes and claude so the tone is classically LLM'd (sales pitch-y is a nice way to put it). I did a few of slop cleanup but I'll do another to tighten it up.

**Comment by ollpu** [comment]
Alright! In that case I'd also cut down on the length of some of the pages especially around the "recommendations" and "design principles". They're reasonable for the most part, but too generic to add much value IMO.

**Comment by systemsoverload** [comment]
Gave it another once over and removed some of the overly flowery blocks. It reads as a reasonable first-time-dev resource to me, but please let me know what else you'd like to see changed. Thanks!

File: guide/src/dev/html-generation.md
Comment (ollpu) [inline guide/src/dev/html-generation.md:114]: Custom writers are not accepted by the public API of the HTML renderer, although the `StrWrite` abstraction is exposed in `pulldown-cmark-escape`. `io/fmt::write` should be sufficient for most use cases like this.

File: guide/src/dev/html-generation.md
Comment (ollpu) [inline guide/src/dev/html-generation.md:161]: 4 & 5 don't make sense to me

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:88]: Untrue - there are multiple passes (two principle ones but also many smaller ones depending on how you look at it)

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:113]: This is not a performance optimization. It is to work around the performance optimization that Text events are often split into multiple parts for various reasons. (Zero-copy ranges may be disjoint & parsed special characters are cheaper to keep in `InlineStr` rather than concatenating into a new allocation.)

You may want to use `TextMergeStream` if you'd rather handle the consecutive events "merged" together, but it has a performance cost.

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:141]:
```suggestion
Several protections are in place to prevent quadratic time or memory usage on malicious input:
```

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:186]: no...

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:169]: I don't think `cows` and `links` are really reused. The main purpose of these is to keep tree nodes small and lifetime-free (index over pointer). The latter is not really necessary, `Tree<Item<'input>>` would work too. Some other fields in `Allocations` are reused I suppose.

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:196]: These would make sense I guess but we don't have them.

File: guide/src/dev/index.md
Comment (ollpu) [inline guide/src/dev/index.md:16]: The whole input string and block structure tree do have to be kept in memory.

File: guide/src/dev/index.md
Comment (ollpu) [inline guide/src/dev/index.md:33]: First and last point are untrue

File: guide/src/dev/index.md
Comment (ollpu) [inline guide/src/dev/index.md:42]: Not really, options only affect parsing ATM

File: guide/src/dev/index.md
Comment (ollpu) [inline guide/src/dev/index.md:44]: This might be a good place to mention the higher level directory structure too, not just the main crate under `pulldown-cmark/`.

**Review by ollpu — COMMENTED** [review]
Tried not being too pedantic, but here we are :slightly_smiling_face:

**Comment by systemsoverload** [comment]
Excellent! Thanks for the feedback, pedantry is always welcome when talking docs. Will take another pass at the branch this morning.

File: guide/src/dev/inline-processing.md
Comment (systemsoverload) [reply guide/src/dev/inline-processing.md]: Ive removed that block, but that feels worth mentioning in the performance and/or extensions guide

File: guide/src/dev/string-handling.md
Comment (systemsoverload) [reply guide/src/dev/string-handling.md:122]: I wanted to call out `TextMergeStream` somewhere in here, but maybe its fine to omit as it is documented pretty extensively elsewhere.

Comment (ollpu) [reply guide/src/dev/string-handling.md:122]: Yeah, I think you could mention the "separate text events" as one optimization, which the user can opt out of with `TextMergeStream` if desired. F

File: guide/src/dev/performance.md
Comment (systemsoverload) [reply guide/src/dev/performance.md:141]: Rewrote to integrate your suggestion WRT emphasizing linear performance:

``` ## Protection Against Pathological Input

It is important that the parser performance remain linear with respect to the input length, otherwise the parser would find itself vulnerable to potential DOS attacks. This may not be important for all consumers, but for anyone depending on the library to handle user generated content this is critical.

Several protections are in place to prevent quadratic time or memory usage on malicious input:```

File: guide/src/dev/index.md
Comment (systemsoverload) [reply guide/src/dev/index.md:42]: This was one of several issues caused by me working (accidentally) on an ancient version of the library for an attempt at HTML parsing improvements. Removed.

File: guide/src/dev/html-generation.md
Comment (systemsoverload) [reply guide/src/dev/html-generation.md:115]: How about something like:

` - Custom writer implementations via implementing the `StrWrite` trait`

This is the bit of the crate that I was personally the most interested int and it wasn't clear to me using the existing docs.

**Comment by systemsoverload** [comment]
Accepted your suggestions, made some more copy edits, and rebased everything together. I only have the one remaining question as to how (if at all) you think we should document the intent of the `StrWrite` trait.

**Comment by ollpu** [comment]
Thanks! I'm not seeing the rebase yet though, and it looks like the latest commit undoes the changes.

> How about something like:
>
> - Custom writer implementations via implementing the StrWrite trait
> 
> This is the bit of the crate that I was personally the most interested int and it wasn't clear to me using the existing docs.

That isn't possible right now though. The only public entry points for the HTML module are:

```rust
pub fn push_html<'a, I>(s: &mut String, iter: I) {...}
pub fn write_html_fmt<'a, I, W: std::fmt::Write>(writer: W, iter: I) {...}
pub fn write_html_io<'a, I, W: std::io::Write>(writer: W, iter: I) {...}
```

`StrWrite` is mainly meant to be an adapter trait. An implementation detail we'd rather not have to expose. If you look at the definition for [`std::fmt::Write`](https://doc.rust-lang.org/stable/core/fmt/trait.Write.html), it is very close to `StrWrite`:

```rust
pub trait Write {
    // Required method
    fn write_str(&mut self, s: &str) -> Result;

    // Provided methods
    fn write_char(&mut self, c: char) -> Result { ... }
    fn write_fmt(&mut self, args: Arguments<'_>) -> Result { ... }
}
```

So for all intents and purposes, a user of pulldown-cmark should be able to implement `fmt::Write` for their custom type and use that instead of `StrWrite`.

The reason we want `StrWrite` at all is it makes the `Error` type customizable as an associated type, so we can pass through both `std::fmt::Result` and `std::io::Result` without losing anything.

Reading the [docs](https://doc.rust-lang.org/std/fmt/struct.Error.html) for `std::fmt::Error`, I notice it mentions that the way IO errors are passed through formatting in the standard library is [with an adapter that stores the error](https://doc.rust-lang.org/src/std/io/mod.rs.html#1821):

```rust
struct Adapter<'a, T: ?Sized + 'a> {
    inner: &'a mut T,
    error: Result<()>,
}
```

We could consider something similar for pulldown-cmark in the future as well.

Previous discussions:
https://github.com/pulldown-cmark/pulldown-cmark/issues/492
https://github.com/pulldown-cmark/pulldown-cmark/pull/493
https://github.com/pulldown-cmark/pulldown-cmark/pull/870
https://github.com/pulldown-cmark/pulldown-cmark/pull/881

TL;DR: `StrWrite` is an implementation detail to abstract over the `fmt` and `io` machinery, since they have different error types. Users can provide custom writers by implementing either `fmt::Write` or `io::Write`, whichever is more appropriate.

**Comment by systemsoverload** [comment]
Unrelated to this PR, that actually cleared up some mystery around the `W` generic 👍

**Comment by ollpu** [comment]
Indeed, I had to read through those prior discussions again to remind myself of the nuances

**Comment by systemsoverload** [comment]
Actually pushed the rebase now ;)

File: guide/src/dev/inline-processing.md
Comment (ollpu) [inline guide/src/dev/inline-processing.md:3]: Forgot this, for completeness
```suggestion
The second pass of pulldown-cmark's parsing process handles inline elements like emphasis, links, and code spans.

Corresponds to Phase 2 of ["A Parsing Strategy"](https://spec.commonmark.org/0.31.2/#appendix-a-parsing-strategy).
```

File: guide/src/dev/performance.md
Comment (ollpu) [inline guide/src/dev/performance.md:17]: There is no simd.rs

**Review by ollpu — APPROVED** [review]
Couple notes still, but I won't hold this up too much more

File: guide/src/dev/inline-processing.md
Comment (ollpu) [reply guide/src/dev/inline-processing.md:3]: Actually, we don't really quite use the described algorithms for phase 2, even though it's conceptually doing the same things. I'll just close this thread

**Comment by Martin1887** [comment]
Thanks for the writing and reviewing work! I will try to review this in the next days.

**Comment by Martin1887** [comment]
It looks good to me! I think it can be merged when the last olipu's comment is fixed. Thanks!

**Comment by systemsoverload** [comment]
Corrected that reference to be `firstpass.rs` 👍

**Pull request merged**
